FWCore/Framework/src/LuminosityBlockProcessingStatus.h

```diff
@@ -103,6 +115,7 @@ namespace edm {
     std::vector<std::shared_ptr<const EventSetupImpl>> eventSetupImpls_;
     WaitingTaskList endIOVWaitingTasks_;
     edm::WaitingTaskHolder globalEndRunHolder_;
+    std::vector<std::atomic<bool>> streamStartedLumi_;
```

Comment:
I think there might be an easier way. It appears to me that there are now two questions that need to be answered
1. should a stream start processing this lumi
2. once a stream ends processing the lumi, is this the last stream to finish that lumi

I think we could answer these questions by adding only one additional new member
```C++
enum class State {kRunning, kUpdating, kDone};
std::atomic<State> state_{State::kRunning};
```

Then having the following two methods
```C++
bool shouldStreamStartLumi() {
        if (state_ == State::kDone) return false;

   	auto expected = State::kRunning;

	bool changed = false;
	do {
	  changed = state_.compare_exchange_strong(expected, State::kUpdating);
	  if (expected == kDone) return false;
	  expected = kRunning;
	} while(changed == false);

	++nStreamsStillProcessingLumi_;
	state_ = State.kRunning;
	return true;
}
```

```C++
bool streamFinishedLumi() {
	auto expected = State.kRunning;
	bool changed = false;
	do {
	  changed = state_.compare_exchange_strong(expected, State::kUpdating);
	  if (expected == State.kUpdating) expected = State.kRunning;
          //NOTE if expected == State.kDone, that is fine so no need to change to kRunning
	} while(changed ==false);
	auto remaining = --nStreamsStillProcessingLum_;
	state_ = State::kDone;
	return 0 == remaining;	
}
```
state_ is protecting `nStreamsStillProcessingLumi_` so only if we are in kRunning state can the value be incremented and therefore it is safe to decrement and check for 0 as once it reaches 0 it can never increase again. Because of this protection, `nStreamsStillProcessingLumi_` would no longer have to be `std::atomic`.

Reply:
I mostly followed Chris's suggestion. I like that it eliminated the vector of atomics with a single atomic. That is better. The logic is a little more complex with the spin locks and such and probably a few more atomic operations, but memory usage is reduced.

Some modifications on top of Chris's suggestion. I added a new function that mostly does what the suggested version of streamFinishedLumi does and named it noMoreEventsInLumi. It is called once per lumi instead of once per stream per lumi. nStreamsStillProcessingLumi_ remains an atomic, but there is a new member called nStreamsProcessingLumi_ used in the incrementing that is not atomic that gets copied at the correct point. And I changed one of the enum names from kDone to kNoMoreEvents.

The changes are in EventProcessor.cc (the .h file is no longer modified) and LuminosityBlockProcessingStatus.h and .cc. That is the most important part. All the rest is related to unit tests. Existing unit tests should all be working now and I am currently thinking about whether new unit tests are needed and what they ought to do.